packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+        },
+        ...templateFunctions,
+      ],
+    };
+
+    const decorator = BindDecoratorFactory.createDecorator(
+      BINDING_METADATA_KEY,
+      spec,
+    );
+    decorator(target);
+  };
+}
+
+export namespace bind {
+  /**
+   * `@bind.provider` to denote a provider class
+   *
+   * A list of binding scope/tags or template functions to configure the binding
+   */
+  export function provider(
+    ...templates: (BindingTemplate | BindingScopeAndTags)[]
+  ): ((target: Constructor<Provider<unknown>>) => void) {
+    return (target: Constructor<Provider<unknown>>) => {
+      bind(
+        // Set up the default for providers
+        binding =>
+          binding.toProvider(target).tag('provider', {type: 'provider'}),
+        // Call other template functions
+        ...templates,
+      )(target);
+    };
+  }
+}
+
+/**
+ * Convert binding scope and tags as a template function
+ * @param scopeAndTags
+ */
+function asBindingTemplate(scopeAndTags: BindingScopeAndTags): BindingTemplate {
+  return binding => {
+    if (scopeAndTags.scope) {
+      binding.inScope(scopeAndTags.scope);
+    }
+    if (scopeAndTags.tags) {
+      if (Array.isArray(scopeAndTags.tags)) {
+        binding.tag(...scopeAndTags.tags);
+      } else {
+        binding.tag(scopeAndTags.tags);
+      }
+    }
+  };
+}
+
+/**
+ * Get binding metadata for a class
+ * @param target The target class
+ */
+export function getBindingMetadata(
+  target: Function,
+): BindingMetadata | undefined {
+  return MetadataInspector.getClassMetadata<BindingMetadata>(
+    BINDING_METADATA_KEY,
+    target,
+  );
+}
+
+/**
+ * Get the binding template from a class with binding metadata
+ *
+ * @param cls A class with optional `@bind`
+ */
+export function bindingTemplateFor(cls: Constructor<unknown>): BindingTemplate {
+  const spec = getBindingMetadata(cls);
+  const templateFunctions = (spec && spec.templates) || [];
+  return binding => {
+    for (const t of templateFunctions) {
+      binding.apply(t);
+    }
+    if (spec && spec.target !== cls) {
+      // Remove name/key from base classes
+      binding.apply(removeNameAndKey);
+    }
+  };
+}
+
+export const TYPE_NAMESPACES: {[name: string]: string} = {
+  controller: 'controllers',
+  repository: 'repositories',
+  model: 'models',
+  dataSource: 'dataSources',
+  server: 'servers',
+  class: 'classes',
+  provider: 'providers',
+};
+
+function getNamespace(type: string) {
+  if (type in TYPE_NAMESPACES) {
+    return TYPE_NAMESPACES[type];
+  } else {
+    return `${type}s`;
+  }
+}
+
+/**
+ * Create a binding from a class with decorated metadata
+ * @param cls A class
+ * @param key Optional binding key. If not present, infer it from the class
+ */
+export function createBindingFromClass<T = unknown>(
+  cls: Constructor<unknown>,
+  key?: string,
+): Binding<T> {
+  const templateFn = bindingTemplateFor(cls);
+  // Create a dummy binding if key is not provided
+  if (!key) {
```

Comment:
Please extract the code building the key into a standalone function and use early returns to keep the indentation level low.

```ts
if (key) return key;
// ...
key = bindingTemplate.tagMap['key'];
if (key) return key;
// ...
key = `${getNamespace(type)}.${name}`;
return key;
```